File: lvl_02/prob_070/prob70.go
package main

import (
	"fmt"
	"math"
	"problem70/SortString"
	"strconv"
)

func phi(n int) int {
	res := float64(n)
	for i := 2; i*i <= n; i++ {
		if n%i == 0 {
			for n%i == 0 {
				n = n / i
			}
			res = res * (1.0 - 1.0/float64(i))
		}
	}
	if n > 1 {
		res = res * (1.0 - 1.0/float64(n))
	}
	return int(res)
}

func main() {
	minRatio := 1000.0
	minPhi := -1
	minIndex := -1
	phiMap := make(map[int]int)
	phiMap[1] = 1
	for i := 0; i < int(math.Pow(10, 7)+1); i++ {
		phiRes := phi(i)
		phiMap[i] = phiRes
		phiSorted := SortString.SortString(strconv.Itoa(phiRes))
		iSorted := SortString.SortString(strconv.Itoa(i))
		if phiRes != i && phiSorted == iSorted {
			currRatio := float64(i) / float64(phiRes)
			if minRatio > currRatio {
				minRatio = currRatio
				minPhi = phiRes
				minIndex = i
				fmt.Printf("%d, %d, %f\n", i, phiRes, float64(i)/float64(phiRes))
			}
		}
	}

	fmt.Printf("%d, %d, %f\n", minIndex, minPhi, minRatio)
}

File: lvl_02/prob_070/prob70.py
'''
Euler's Totient function, φ(n) [sometimes called the phi function], 
is used to determine the number of positive numbers less than or equal to n which are relatively prime to n. 
For example, as 1, 2, 4, 5, 7, and 8, are all less than nine and relatively prime to nine, φ(9)=6.
The number 1 is considered to be relatively prime to every positive number, so φ(1)=1.

Interestingly, φ(87109)=79180, and it can be seen that 87109 is a permutation of 79180.

Find the value of n, 1 < n < 10^7, for which φ(n) is a permutation of n and the ratio n/φ(n) produces a minimum.
'''
# takes too long to iterate and calculate each value
# phi(n)/n == phi(rad(n))/rad(n)
# I think the algorithm for finding rad(n) is less efficient and just adds unnecessary time
# also due to the fact that the only time it was a permutation and lower ratio was when rad(n) = n

def phi(n): 
    result = n
    p = 2
    while p * p <= n : 
        if n % p == 0 : 
            while n % p == 0 : 
                n = n // p 
            result = result * (1.0 - (1.0 / float(p))) 
        p = p + 1
    if n > 1 :
        result = result * (1.0 - (1.0 / float(n))) 
    return int(result) 

# answer is 8319823
# found by going backwards, took longer than acceptable though
# there must be some trick
# I thought that as n approached infinity, that the ratio would also approach a minimum
def main():
    minRatio = 1000
    minPhi = -1
    minIndex = -1
    phiDict = {}
    phiDict[1] = 1
    for x in range(10**7+1, -1, -1):
        '''
        if x % 2 == 0:
            k = x / 2
            phiDict[x] = phiDict[k]
            if k % 2 == 0:
                phiDict[x] = phiDict[x] * 2
            phiRes = phiDict[x]
        else:
        '''
        phiRes = phi(x)
        phiDict[x] = phiRes
        phiSorted = str(sorted(list(str(phiRes))))
        xSorted = str(sorted(list(str(x))))
        if phiRes != x and phiSorted == xSorted:
            currRatio = x/phiRes
            if minRatio > currRatio:
                minRatio = currRatio
                minPhi = phiRes
                minIndex = x
                print(x, phiRes, x/phiRes)

    return

if __name__ == '__main__':
    import time
    start = time.time()
    try:
        main()
    finally:
        print("time:",time.time()-start)